# **Lab 1: ESIEECom - Campagnes de publicité**

In [ ]:
! pip install pandas

In [ ]:
import pandas as pd

# Enoncé

Vous travaillez au sein de **ESIEECom**, une entreprise de communication qui a pour objectif de créer des campagnes de publicité pour des entreprises tierces, dans le but d'inciter les gens à acheter des produits chez ces dîtes entreprises. Ainsi, plusieurs campagnes marketing ont été diffusées à 29 758 clients potentiels (une seule campagne par client) pour le compte d'une entreprise marketing **greenSCart**.


Plusieurs semaines sont passées, il est à présent temps d'analyser les performances des campagnes en termes de clics et d'achats. Votre rôle dans cette analyse va être de préparer le terrain, en formattant les données pour les rendre exploitables.

Vous avez pour cela trois fichiers à votre disposition :
- *campaign.csv* : un fichier présentant, par client, la campagne reçue et le canal d'acquisition (external_site_id)
- *click.csv* : un fichier présentant l'ensemble des clients ayant cliqué au moins une fois sur le lien contenu dans la campagne, avec la date du premier clic
- *purchase.csv* : un fichier présentant l'ensemble des clients ayant acheté parmi ceux ayant cliqué sur le lien, avec la date d'achat et une série d'informations sur le client.


# Exercice

## Partie 1 : analyse sur les campagnes

### 1. Importation 

Vous commencerez par importer le fichier campaign.csv. Pour cela, vous utiliserez la fonction [`read_csv`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html).

In [ ]:
campaign = pd.read_csv('campaign.csv')

In [ ]:
campaign

client_id  campaign_id  canal_id
0      2392769782            3    370668
1      2693949047            1    339897
2      6841408721            2    989339
3      1457891936            1    530191
4      5354259272            1    206900
...           ...          ...       ...
29753  2701723551            3    140514
29754  7271996385            1    892234
29755  4035966889            1    563313
29756  7499403392            1    325552
29757  2017465235            1    812611

[29758 rows x 3 columns]

La première étape après toute importation est de regarder les premières lignes, cela permet :
- de vérifier que l'importation s'est bien passée
- d'avoir un apperçu rapide de ce à quoi ressemble notre dataframe

In [ ]:
campaign.sample(10)

client_id  campaign_id  external_site_id
12125  3028638661            1            804952
21989  1104487119            1            424087
2872     52510814            1            561239
22981  7068175889            3            380034
23206  9592769884            2            656637
28536   610343808            1            958665
17730  1524702166            3            642614
27515  6211825211            3            886881
8804   4232560670            2            572492
24551  8778803062            1            631219

La seconde chose à faire est une petite vérification sur les types de chaque colonne attribués par pandas lors de l'importation, encore une fois pour vérifier que l'importation s'est bien déroulée. Affichez le type de l'ensemble des colonnes

In [ ]:
campaign.dtypes

client_id           int64
campaign_id         int64
external_site_id    int64
dtype: object

Combien de lignes et colonnes sont contenues dans notre fichier ?

In [ ]:
campaign.shape

(29758, 3)

Modifiez le nom de la variable `external_site_id` par `canal_id`.

In [ ]:
campaign= campaign.rename(   columns={ "external_site_id" : "canal_id"}   )

In [ ]:
campaign

client_id  campaign_id  canal_id
0      2392769782            3    370668
1      2693949047            1    339897
2      6841408721            2    989339
3      1457891936            1    530191
4      5354259272            1    206900
...           ...          ...       ...
29753  2701723551            3    140514
29754  7271996385            1    892234
29755  4035966889            1    563313
29756  7499403392            1    325552
29757  2017465235            1    812611

[29758 rows x 3 columns]

### 2. Méthodes de series

Il va être temps de commencer notre analyse en déterminant le nombre de campagnes différentes qui ont été envoyées. En premier lieu, sélectionnez uniquement la colonne `campaign_id`.

In [ ]:
campaign.campaign_id

0        3
1        1
2        2
3        1
4        1
        ..
29753    3
29754    1
29755    1
29756    1
29757    1
Name: campaign_id, Length: 29758, dtype: int64

Même si cette dernière est codée avec des nombres, il s'agit bien d'une variable **qualitative**. Nous verrons un peu plus tard comment changer le type, mais nous chercherons avant tout le nombre de valeurs différentes au sein de notre variable : trouvez une [méthode](https://pandas.pydata.org/docs/reference/api/pandas.unique.html) appliquée à cette colonne permettant d'obtenir cela.

In [ ]:
campaign['campaign_id'].unique()

array([3, 1, 2])

De façon assez similaire, combien de canaux d'acquisition (`canal_id`) différents ont été utilisés ?

In [ ]:
len(campaign['canal_id'].unique())

12880

Nous souhaitons à présent avoir le nombre de personnes qui ont été contactées par campagne, en gardant bien en tête qu'un individu n'a pu recevoir qu'une seule campagne sur les trois envoyées, donc une ligne de notre dataframe correspond à un client unique

In [ ]:
campaign['campaign_id'].value_counts()

1    10038
3     9863
2     9857
Name: campaign_id, dtype: int64

### 3. Ajout des informations de click et de purchase

Dans un premier temps, importez les fichiers *click.csv* et *purchase.csv*, et affichez les 5 premières lignes

In [ ]:
click= pd.read_csv('click.csv')

Combien de cliques en total ?

In [ ]:
click.shape

(1775, 2)

In [ ]:
click

click_date   client_id
0     2025-11-01 01:57:47  4217174943
1     2025-11-01 02:11:21  8581188253
2     2025-11-01 02:13:16  2693232025
3     2025-11-01 03:28:12  8905731989
4     2025-11-01 03:51:23  2071156755
...                   ...         ...
1770  2025-12-01 10:29:52  4007461204
1771  2025-12-01 10:37:34  4493292524
1772  2025-12-01 10:38:14  5302443502
1773  2025-12-01 10:41:52  7295507401
1774  2025-12-01 10:42:05  2965986483

[1775 rows x 2 columns]

Verifier que nous avions bien un seul clique par client !

In [ ]:
len(click['client_id'].unique())

1775

In [ ]:
purchase= pd.read_csv('purchase.csv')

In [ ]:
purchase

purchase_date   client_id  product_id gender  price  age
0       1763303606  7048447933           1      m    354   65
1       1763341126  9138057668           8      f    165   37
2       1763398855  1327443078           3      f     29   28
3       1763455517  1083555113           3      f     29   23
4       1763543415   665955021           6      f     34   25
..             ...         ...         ...    ...    ...  ...
135     1766515963    67211265           3      m     29   27
136     1766531026  6200686769           3      f     29   21
137     1766638894  7847011933           3      f     29   19
138     1766662814  7896338445           8      m    165   47
139     1766690086  2066294216           5      f     89   45

[140 rows x 6 columns]

Il va falloir à présent ajouter les informations de click et purchase à notre dataframe initial campaign. Nous pourrons faire cela via la fonction [`merge`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html) de pandas.

Il existe 3 grands type de [`jointures`](https://medium.com/@yannicknoubissi2018/sql-jointures-eefc6e77c85d) :
- interne
- externe
- à droite ou à gauche

Ici, on souhaite absolument conserver l'ensemble des informations qui sont dans notre dataframe initial (campaign). Faites dans un premier temps la jointure adaptée entre *campaign* et *click* :

In [ ]:
 df_final= pd.merge(campaign, click, on='client_id', how='left')

In [ ]:
df_final

client_id  campaign_id  canal_id click_date
0      2392769782            3    370668        NaN
1      2693949047            1    339897        NaN
2      6841408721            2    989339        NaN
3      1457891936            1    530191        NaN
4      5354259272            1    206900        NaN
...           ...          ...       ...        ...
29753  2701723551            3    140514        NaN
29754  7271996385            1    892234        NaN
29755  4035966889            1    563313        NaN
29756  7499403392            1    325552        NaN
29757  2017465235            1    812611        NaN

[29758 rows x 4 columns]

Vous avez normalement des NaN qui apparaissent : pas de panique tout est parfaitement normal, car sur l'ensemble des clients, tous n'ont pas cliqué !

Faites ensuite la jointure entre le dataframe obtenu et la table purchase, de la même façon. Vous nommerez le dataframe final, `df_final` :

In [ ]:
df_final= pd.merge(df_final, purchase, on='client_id', how='left')

In [ ]:
df_final

client_id  campaign_id  canal_id  ... gender  price  age
0      2392769782            3    370668  ...    NaN    NaN  NaN
1      2693949047            1    339897  ...    NaN    NaN  NaN
2      6841408721            2    989339  ...    NaN    NaN  NaN
3      1457891936            1    530191  ...    NaN    NaN  NaN
4      5354259272            1    206900  ...    NaN    NaN  NaN
...           ...          ...       ...  ...    ...    ...  ...
29753  2701723551            3    140514  ...    NaN    NaN  NaN
29754  7271996385            1    892234  ...    NaN    NaN  NaN
29755  4035966889            1    563313  ...    NaN    NaN  NaN
29756  7499403392            1    325552  ...    NaN    NaN  NaN
29757  2017465235            1    812611  ...    NaN    NaN  NaN

[29758 rows x 9 columns]

### 4. Analyse rapide des informations de notre table

Utilisez les fonctions [`describe`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) et [`info`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html) pour avoir quelques informations rapides sur les variables de notre dataframe :

In [ ]:
df_final.describe()

client_id   campaign_id  ...       price         age
count  2.975800e+04  29758.000000  ...  140.000000  140.000000
mean   5.026531e+09      1.994119  ...  145.807143   39.085714
std    2.905848e+09      0.817771  ...  112.746478   15.167879
min    3.544420e+05      1.000000  ...   29.000000   19.000000
25%    2.520960e+09      1.000000  ...   67.000000   27.000000
50%    5.042014e+09      2.000000  ...   89.000000   34.500000
75%    7.548261e+09      3.000000  ...  299.000000   56.000000
max    9.999882e+09      3.000000  ...  354.000000   69.000000

[8 rows x 7 columns]

In [ ]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29758 entries, 0 to 29757
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   client_id      29758 non-null  int64  
 1   campaign_id    29758 non-null  int64  
 2   canal_id       29758 non-null  int64  
 3   click_date     1775 non-null   object 
 4   purchase_date  140 non-null    float64
 5   product_id     140 non-null    float64
 6   gender         140 non-null    object 
 7   price          140 non-null    float64
 8   age            140 non-null    float64
dtypes: float64(4), int64(3), object(2)
memory usage: 2.3+ MB


A partir de cette dernière méthode, pouvez-vous déterminer le nombre de personnes ayant cliqué et le nombre de personnes ayant acheté parmi ces dernières ?

### 5. Changement de type

En regardant le résultat de la méthode *info*, on se rend compte que plusieurs variables n'ont pas le bon type, notamment : client_id, campaign_id, canal_id et product_id qui ont été importées en int ou float, alors que ce sont des variables qualitatives.

[Modifiez](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.astype.html) donc leur type pour les mettre en string ou category.

In [ ]:
var_modif=['client_id', 'campaign_id','canal_id', 'product_id']

for var in var_modif:
  df_final[var]= df_final[var].astype('category')


On remarque également que click_date et purchase_date n'ont pas le bon format : ce sont censés être des dates, alors qu'elles sont actuellement codées comme étant respectivement chaine de caractère (`object`) et décimaux (`float`). La fonction utile dans ce cas sera [`to_datetime`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html)

Convertissez click_date au format datetime :

In [ ]:
df_final['click_date']= pd.to_datetime (  df_final['click_date'], format= '%Y-%m-%d %H:%M:%S'  )

In [ ]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 29758 entries, 0 to 29757
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   client_id      29758 non-null  category      
 1   campaign_id    29758 non-null  category      
 2   canal_id       29758 non-null  category      
 3   click_date     1775 non-null   datetime64[ns]
 4   purchase_date  140 non-null    float64       
 5   product_id     140 non-null    category      
 6   gender         140 non-null    object        
 7   price          140 non-null    float64       
 8   age            140 non-null    float64       
dtypes: category(4), datetime64[ns](1), float64(3), object(1)
memory usage: 3.7+ MB


*purchase_date* est un peu particulière, car le format est ce qu'on appelle un timestamp. Le timestamp désigne le nombre de secondes écoulées depuis le 1er janvier 1970 à minuit UTC précises jusqu'à la date exprimée. Même si le format diffère, la fonction de conversion reste la même, il suffira de modifier les options de cette dernière.

In [ ]:
df_final['purchase_date']= pd.to_datetime (  df_final['purchase_date'], unit= 's'  )

In [ ]:
df_final

client_id campaign_id canal_id click_date  ... product_id gender price  age
0      2392769782           3   370668        NaT  ...        NaN    NaN   NaN  NaN
1      2693949047           1   339897        NaT  ...        NaN    NaN   NaN  NaN
2      6841408721           2   989339        NaT  ...        NaN    NaN   NaN  NaN
3      1457891936           1   530191        NaT  ...        NaN    NaN   NaN  NaN
4      5354259272           1   206900        NaT  ...        NaN    NaN   NaN  NaN
...           ...         ...      ...        ...  ...        ...    ...   ...  ...
29753  2701723551           3   140514        NaT  ...        NaN    NaN   NaN  NaN
29754  7271996385           1   892234        NaT  ...        NaN    NaN   NaN  NaN
29755  4035966889           1   563313        NaT  ...        NaN    NaN   NaN  NaN
29756  7499403392           1   325552        NaT  ...        NaN    NaN   NaN  NaN
29757  2017465235           1   812611        NaT  ...        NaN    NaN   NaN  NaN

[29758 rows x 9 columns]

In [ ]:
df_final.dtypes

client_id              category
campaign_id            category
canal_id               category
click_date       datetime64[ns]
purchase_date    datetime64[ns]
product_id             category
gender                   object
price                   float64
age                     float64
dtype: object

### 6. la sélection au sein d'un dataframe

Nous avons d'ores et déjà vu comment sélectionner une colonne. En utilisant une écriture similaire, sélectionnez deux colonnes de votre choix.

In [ ]:
df_final[  [ 'client_id', 'campaign_id']    ]

client_id campaign_id
0      2392769782           3
1      2693949047           1
2      6841408721           2
3      1457891936           1
4      5354259272           1
...           ...         ...
29753  2701723551           3
29754  7271996385           1
29755  4035966889           1
29756  7499403392           1
29757  2017465235           1

[29758 rows x 2 columns]

Il est à présent temps d'utiliser la fonction de sélection qui vous sera le plus utile, la méthode [`loc`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html). A partir de cette dernière, sélectionnez et déterminez le nombre de clients dont on a l'information, qui ont plus de 65 ans :

In [ ]:
df_final.loc[  df_final['age']>65    ,  :]

client_id campaign_id canal_id  ... gender  price   age
552    8751879875           1   429717  ...      f  354.0  68.0
3109   5031077513           1   201592  ...      f  165.0  67.0
7840   3755043429           2   335839  ...      f  354.0  68.0
11728  7928559147           1   196935  ...      f  354.0  66.0
12758  3179544440           1   970704  ...      f  354.0  69.0

[5 rows x 9 columns]

Sélectionnez uniquement les identifiants de ces clients :

In [ ]:
df_final.loc[  df_final['age']>65    ,  'client_id' ]

552      8751879875
3109     5031077513
7840     3755043429
11728    7928559147
12758    3179544440
Name: client_id, dtype: category
Categories (29758, int64): [354442, 522647, 668537, 692084, ..., 9998833142, 9998909223, 9999085476,
                            9999881734]

Sélectionnez à présent les clients de moins de 25 ans ayant acheté le produit 8 :

In [ ]:
df_final.loc[ ( df_final['age']<25 ) & (df_final['product_id'] ==8  )  ,  'client_id']

548      7331215946
18629    9353335552
Name: client_id, dtype: category
Categories (29758, int64): [354442, 522647, 668537, 692084, ..., 9998833142, 9998909223, 9999085476,
                            9999881734]

Sur ces clients, sélectionnez uniquement leur identifiant et l'identifant de la campagne qu'ils ont reçue

In [ ]:
df_final.loc[ ( df_final['age']<25 ) & (df_final['product_id'] ==8  )  ,  ['client_id', 'campaign_id' ]]

client_id campaign_id
548    7331215946           2
18629  9353335552           3

Pour finir, vous créerez un nouveau dataframe, qui contiendra uniquement les 140 lignes qui correspondent à un achat client

In [ ]:
achat = df_final.loc[  df_final['purchase_date'].notnull()  ,  :]

Au sein de ce nouveau dataframe, sélectionnez les identifiants des clients ayant reçu la campagne selon l'un des canaux d'acquisition suivant : 

550488, 154492, 721316, 464778, 850899

In [57]:
canaux= [550488, 154492, 721316, 464778, 850899]
achat .loc[achat['canal_id'].isin(canaux), : ]

client_id campaign_id canal_id  ... gender  price   age
9409   3670039263           1   550488  ...      f  109.0  38.0
9770    675653663           1   154492  ...      f   89.0  39.0
10025  8300306275           2   721316  ...      f  299.0  62.0
10797  9432221904           2   464778  ...      m  165.0  36.0
11092  3745071281           2   850899  ...      m  299.0  54.0

[5 rows x 9 columns]

## Partie 2 : analyse sur les achats

Pour la suite de l'exercice, on se concentrera uniquement sur le dernier dataframe créé contenant les informations d'achat

### 1. Création de nouvelles variables

Le prix actuellement indiqué correspond au prix hors taxe. Créez une nouvelle variable correspondant au prix TTC de chaque produit vendu, en sachant que :

$$Prix  TTC = 1.2 * Prix  HT$$

In [58]:
achat['price_ttc']=achat['price']*1.2

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [59]:
achat

client_id campaign_id canal_id  ...  price   age price_ttc
11      665955022           2   646537  ...  354.0  63.0     424.8
51     2693232025           2   673707  ...  299.0  56.0     358.8
87     7847011933           2   276236  ...   29.0  19.0      34.8
174    7048447933           3   182313  ...  354.0  65.0     424.8
302    9554597531           3   341237  ...   89.0  37.0     106.8
...           ...         ...      ...  ...    ...   ...       ...
29134  4651306224           3   620146  ...  299.0  57.0     358.8
29250  2877464022           2   349353  ...  354.0  63.0     424.8
29588  6200686769           2   382063  ...   29.0  21.0      34.8
29622  2885904311           3   865109  ...   29.0  22.0      34.8
29717  4493292524           1   609260  ...   67.0  28.0      80.4

[140 rows x 10 columns]

Il serait également intéressant d'extraire uniquement le jour de l'achat (au format `date`), à partir de la date d'achat pour éventuellement calculer le chiffre d'affaire total par jour.

In [60]:
achat['purchase_day']=achat['purchase_date'].dt.date

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [64]:
!pip list -v | grep [Pp]andas

pandas                        1.1.5          /usr/local/lib/python3.7/dist-packages pip
pandas-datareader             0.9.0          /usr/local/lib/python3.7/dist-packages pip
pandas-gbq                    0.13.3         /usr/local/lib/python3.7/dist-packages pip
pandas-profiling              1.4.1          /usr/local/lib/python3.7/dist-packages pip
sklearn-pandas                1.8.0          /usr/local/lib/python3.7/dist-packages pip


In [61]:
achat

client_id campaign_id canal_id  ...   age price_ttc purchase_day
11      665955022           2   646537  ...  63.0     424.8   2025-11-21
51     2693232025           2   673707  ...  56.0     358.8   2025-11-27
87     7847011933           2   276236  ...  19.0      34.8   2025-12-25
174    7048447933           3   182313  ...  65.0     424.8   2025-11-16
302    9554597531           3   341237  ...  37.0     106.8   2025-12-01
...           ...         ...      ...  ...   ...       ...          ...
29134  4651306224           3   620146  ...  57.0     358.8   2025-12-21
29250  2877464022           2   349353  ...  63.0     424.8   2025-12-22
29588  6200686769           2   382063  ...  21.0      34.8   2025-12-23
29622  2885904311           3   865109  ...  22.0      34.8   2025-12-20
29717  4493292524           1   609260  ...  28.0      80.4   2025-12-16

[140 rows x 11 columns]

### 2. agrégation

A partir de cette variable nouvellement créée et de la fonction [`groupby`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html), créez un nouveau dataframe comportant le chiffre d'affaire TTC total quotidien.

In [65]:
achat.groupby('purchase_day').sum()[['price_ttc']]

price_ttc
purchase_day           
2025-11-16        424.8
2025-11-17        232.8
2025-11-18         34.8
2025-11-19         40.8
2025-11-20        106.8
2025-11-21        424.8
2025-11-22        106.8
2025-11-23        393.6
2025-11-24        729.6
2025-11-25        848.4
2025-11-26        318.0
2025-11-27        519.6
2025-11-28        945.6
2025-11-29       1512.0
2025-11-30       1219.2
2025-12-01       1057.2
2025-12-02        585.6
2025-12-03        970.8
2025-12-04       1096.8
2025-12-05        733.2
2025-12-06        632.4
2025-12-07        540.0
2025-12-08       1344.0
2025-12-09        931.2
2025-12-10        596.4
2025-12-11        459.6
2025-12-12        474.0
2025-12-13        318.0
2025-12-14       1194.0
2025-12-15       1146.0
2025-12-16       1323.6
2025-12-17        420.0
2025-12-18        358.8
2025-12-19        115.2
2025-12-20        232.8
2025-12-21        514.8
2025-12-22        997.2
2025-12-23        256.8
2025-12-25        339.6

# Fin !